File: lib/spells/effects/Heal.cpp
/*
 * Heal.cpp, part of VCMI engine
 *
 * Authors: listed in file AUTHORS in main folder
 *
 * License: GNU General Public License v2.0 or later
 * Full text of license available in license.txt file, in main folder
 *
 */
#include "StdInc.h"

#include "Heal.h"
#include "Registry.h"
#include "../ISpellMechanics.h"

#include "../../battle/IBattleState.h"
#include "../../battle/CUnitState.h"
#include "../../battle/CBattleInfoCallback.h"
#include "../../battle/Unit.h"
#include "../../networkPacks/PacksForClientBattle.h"
#include "../../serializer/JsonSerializeFormat.h"

VCMI_LIB_NAMESPACE_BEGIN

namespace spells
{
namespace effects
{

void Heal::apply(ServerCallback * server, const Mechanics * m, const EffectTarget & target) const
{
	apply(m->getEffectValue(), server, m, target);
}

SpellEffectValue Heal::getHealthChange(const Mechanics * m, const EffectTarget & spellTarget) const
{
	SpellEffectValue result;

	for(const auto & oneTarget : spellTarget)
	{
		const battle::Unit * unit = oneTarget.unitValue;

		if(unit)
		{
			result += getHealEffectValue(m->getEffectValue(), m, unit);
			result.unitType = unit->creatureId();
		}
	}
	return result;
}

void Heal::apply(int64_t value, ServerCallback * server, const Mechanics * m, const EffectTarget & target) const
{
	BattleLogMessage logMessage;
	logMessage.battleID = m->battle()->getBattle()->getBattleID();

	BattleUnitsChanged pack;
	pack.battleID = m->battle()->getBattle()->getBattleID();

	prepareHealEffect(value, pack, logMessage, *server->getRNG(), m, target);
	if(!pack.changedStacks.empty())
		server->apply(pack);
	if(!logMessage.lines.empty())
		server->apply(logMessage);
}

bool Heal::isValidTarget(const Mechanics * m, const battle::Unit * unit) const
{
	const bool onlyAlive = healLevel == EHealLevel::HEAL;
	const bool validInGenaral = unit->isValidTarget(!onlyAlive);

	if(!validInGenaral)
		return false;

	auto injuries = unit->getTotalHealth() - unit->getAvailableHealth();

	if(injuries == 0)
		return false;

	if(minFullUnits > 0)
	{
		auto hpGained = std::min(m->getEffectValue(), injuries);
		if(hpGained < minFullUnits * unit->getMaxHealth())
			return false;
	}

	if(unit->isDead())
	{
		//check if alive unit blocks resurrection
		for(const BattleHex & hex : unit->getHexes())
		{
			auto blocking = m->battle()->battleGetUnitsIf([hex, unit](const battle::Unit * other)
			{
				return other->isValidTarget(false) && other->coversPos(hex) && other != unit;
			});

			if(!blocking.empty())
				return false;
		}
	}
	return true;
}

void Heal::serializeJsonUnitEffect(JsonSerializeFormat & handler)
{
	static const std::vector<std::string> HEAL_LEVEL_MAP =
	{
		"heal",
		"resurrect",
		"overHeal"
	};

	static const std::vector<std::string> HEAL_POWER_MAP =
	{
		"oneBattle",
		"permanent"
	};

	handler.serializeEnum("healLevel", healLevel, EHealLevel::HEAL, HEAL_LEVEL_MAP);
	handler.serializeEnum("healPower", healPower, EHealPower::PERMANENT, HEAL_POWER_MAP);
	handler.serializeInt("minFullUnits", minFullUnits);
}

SpellEffectValue Heal::getHealEffectValue(int64_t value, const Mechanics * m, const battle::Unit * unit, std::shared_ptr<battle::Unit> newState) const
{
	SpellEffectValue result {};

	result.hpDelta = value;

	if(!newState)
		newState = unit->acquire();
	const auto countBeforeHeal = newState->getCount();
	newState->heal(result.hpDelta, healLevel, healPower);

	result.unitsDelta = std::max(0, newState->getCount() - countBeforeHeal);

	return result;
}

void Heal::prepareHealEffect(int64_t value, BattleUnitsChanged & pack, BattleLogMessage & logMessage, RNG & rng, const Mechanics * m, const EffectTarget & target) const
{
	for(const auto & oneTarget : target)
	{
		const battle::Unit * unit = oneTarget.unitValue;

		if(unit)
		{
			auto state = unit->acquire();
			SpellEffectValue healValue = getHealEffectValue(value, m, unit, state);

			if(healValue.hpDelta > 0)
			{
				if(healValue.unitsDelta > 0)
				{
					// %d %s rise from the dead!
					// in the table first comes plural string, then the singular one
					MetaString resurrectText;
					state->addText(resurrectText, EMetaText::GENERAL_TXT, 116, healValue.unitsDelta == 1);
					state->addNameReplacement(resurrectText);
					resurrectText.replaceNumber(healValue.unitsDelta);
					logMessage.lines.push_back(std::move(resurrectText));
				}
				else if (healValue.hpDelta > 0 && m->caster->getHeroCaster() == nullptr) //Show text about healed HP if healed by unit
				{
					MetaString healText;
					auto casterUnitID = m->caster->getCasterUnitId();
					auto casterUnit = m->battle()->battleGetUnitByID(casterUnitID);
					healText.appendLocalString(EMetaText::GENERAL_TXT, 414);
					casterUnit->addNameReplacement(healText, false);
					state->addNameReplacement(healText, false);
					healText.replaceNumber((int)healValue.hpDelta);
					logMessage.lines.push_back(std::move(healText));
				}

				if(healValue.hpDelta > 0)
				{
					UnitChanges info(state->unitId(), UnitChanges::EOperation::RESET_STATE);
					info.healthDelta = healValue.hpDelta;
					state->save(info.data);
					pack.changedStacks.push_back(info);
				}
			}
		}
	}
}


}
}

VCMI_LIB_NAMESPACE_END

File: scripts/heal.lua
local Base = require("unitEffect")
local Script = setmetatable({}, {__index = Base})
Script.__index = Script

local HEAL_LEVEL_FROM_STRING = {
	heal      = ENUM.HealLevel.heal,
	resurrect = ENUM.HealLevel.resurrect,
	overHeal  = ENUM.HealLevel.overheal,
}
local HEAL_POWER_FROM_STRING = {
	oneBattle = ENUM.HealPower.oneBattle,
	permanent = ENUM.HealPower.permanent,
}

function Script:getHealLevel()
	return HEAL_LEVEL_FROM_STRING[self.healLevel] or ENUM.HealLevel.heal
end
function Script:getHealPower()
	return HEAL_POWER_FROM_STRING[self.healPower] or ENUM.HealPower.permanent
end
function Script:getMinFullUnits()
	return self.minFullUnits or 0
end

--- Only injured units are valid; resurrect/overheal levels also accept dead units.
function Script:isValidTarget(mechanics, unit)
	local level = self:getHealLevel()
	local allowDead = level ~= ENUM.HealLevel.heal
	if not unit:isValidTarget(allowDead) then return false end

	local injuries = unit:getTotalHealth() - unit:getAvailableHealth()
	if injuries <= 0 then return false end

	local mfu = self:getMinFullUnits()
	if mfu > 0 then
		local hpGained = math.min(mechanics:getEffectValue(), injuries)
		if hpGained < mfu * unit:getMaxHealth() then return false end
	end

	if unit:isDead() then
		local hexes = unit:getHexes()
		for i = 1, hexes:size() do
			local hex = hexes:at(i)
			local blockers = mechanics:getBattle():getUnitsIf(function(other)
				return other ~= unit and other:isValidTarget(false) and other:coversPos(hex)
			end)
			if #blockers > 0 then return false end
		end
	end

	return true
end

--- Returns HP and unit count change for hover tooltip.
function Script:getHealthChange(mechanics, spellTarget)
	local result = { hpDelta = 0, unitsDelta = 0 }
	for _, dest in ipairs(spellTarget) do
		local unit = dest.unit
		if unit then
			local copy = unit:copy()
			local healedHP, resurrected = copy:heal(mechanics:getEffectValue(), self:getHealLevel(), self:getHealPower())
			result.hpDelta   = result.hpDelta   + healedHP
			result.unitsDelta = result.unitsDelta + resurrected
			result.unitType  = unit:getCreature()
		end
	end
	return result
end

--- Heals each target unit and emits battle log messages.
function Script:apply(mechanics, server, target)
	local battle       = mechanics:getBattle()
	local isUnitCaster = mechanics:getHeroCaster() == nil

	for _, dest in ipairs(target) do
		local unit = dest.unit
		if unit then
			local healedHP, resurrected = server:healUnit(
				battle, unit, mechanics:getEffectValue(), self:getHealLevel(), self:getHealPower())

			if resurrected > 0 then
				local textID = resurrected == 1 and "core.genrltxt.117" or "core.genrltxt.116"
				local nameTextID = unit:getCreature():getNameTextID(unit:getCount())
				server:appendLog(battle, {
					append         = { textID },
					replaceStrings = { nameTextID },
					replaceNumbers = { resurrected }
				})
			elseif healedHP > 0 and isUnitCaster then
				local casterUnit = mechanics:getUnitCaster()
				server:appendLog(battle, {
					append         = { "core.genrltxt.414" },
					replaceStrings = {
						casterUnit:getCreature():getNameTextID(1),
						unit:getCreature():getNameTextID(1)
					},
					replaceNumbers = { healedHP }
				})
			end
		end
	end
end

return Script
